core/sr-primitives/src/weights.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Primitives for transaction weighting.
+//!
+//! Each dispatch function withing `decl_module!` can now have an optional
+//! `#[weight = $x]` attribute. $x can be any object that implements the
+//! [`WeighableTransaction`] trait. By default, All transactions are annotated by
+//! `#[weight = TransactionWeight::default()]`.
+//!
+//! Note that the decl_module macro _cannot_ enforce this and will simply fail
+//! if an invalid struct is passed in.
+//!
+//! Note that [`WeighableCall`] and [`WeighableTransaction`] are more or less similar.
+//! The distinction is because one serves to pass the weight from the the
+//! dispatchable function's attribute to the call enum ([`WeighableTransaction`]) and the
+//! other to pass the final weight from call enum to the executive module
+//! ([`WeighableCall`]).
+
+/// The final type that each `#[weight = $x:expr]`'s
+/// expression must evaluate to. Consists of `(base_weight, byte_weight)`.
+pub type Weight = (u32, u32);
+
+// TODO #2431 this value is now set in a way that the weight in the transaction
+// is just the size of input. Must be updated based based on benchmarks. Same for
+// maximum_weight and ::max() variant of TransactionWeight.
+/// The default weight as literal value.
+pub const DEFAULT_WEIGHT: Weight = (0, 1);
+
+/// A `Call` enum that can be weighted using the custom weight attribute of the
+/// its dispatchable functions. Is implemented by default in the `decl_module!`.
+pub trait WeighableCall {
+	/// Return the weight of this call.
+	fn weight(&self) -> Weight;
```

Comment:
Why not pass `encoded_len` as parameter? It would allow strictly more cases to be handled. I know that currently we always have `base_w + bytes * byte_w`, but passing the encoded bytes len would be more future proof.

Reply:
What would you do with the encoded len? Would we need to pass this then down to the actual calculation happening in the `weight` attribute?

Reply:
Had a talk with @pepyakin and concluded: most generic is to actually pass `Call` into this function that the user can provide. Agreed? Bit of an overdue w.r.t. what the Polkadot goal is (Polkadot does not really care about this stuff and wants these `[#weight = X]` to be static values). But smart contract platform can befinit from them quite a lot and state the weight based on arguments, origin, etc.

Reply:
You mean pass the arguments of the transaction? As the user is probably not interested in the `Call` enum instance.

Reply:
and I should add that I think all of this discussion is about the other trait, `WeighableTransaction`. This one afaik will stay like this and is only used so that `executive` can call sth like `Call::weight()` for some particular transaction's call. Explained here 
https://github.com/paritytech/substrate/pull/2799/files#diff-74eb2b089bc101850d8e1b995695c0d6R27